File: Metro_Card/src/station_builder.py
from .base import BaseStation, BaseToken
from .constants import OutputPrefix, PassengerCharges
from typing import List


class MetroStation(BaseStation):

    def __init__(self, name: str):
        super().__init__(name)

    def addPassenger(self, token: BaseToken, passenger_type: PassengerCharges, discount: int):
        return super().addPassenger(token, passenger_type, discount)

    def printSummary(self):
        print(OutputPrefix.TOTAL_COLLECTION.value,
              self.getName(), self.getEarnings(), self.getDiscountsGiven())
        print(OutputPrefix.PASSENGER_TYPE_SUMMARY.value)
        for passenger in self.getPassengersList():
            print(passenger[0], passenger[1])


class StationBuilder:

    def __init__(self):
        self._station_pool = dict()

    def addStation(self, station: BaseStation) -> bool:
        if station.getName() in self._station_pool:
            return False
        self._station_pool[station.getName()] = station
        return True

    def getStation(self, station_name: str) -> BaseStation:
        return self._station_pool.get(station_name)

    def getAllStations(self) -> List[BaseStation]:
        return [station for station in self._station_pool.values()]

File: Metro_Card_Node/src/station_builder.js
const {
    BaseStation
} = require('./base')
const {
    OutputPrefix
} = require('./constants')


class MetroStation extends BaseStation {
    constructor(name) {
        super(name)
    }

    addPassenger = function(token, passenger_type, discount) {
        return super.addPassenger(token, passenger_type, discount)
    }

    printSummary = function() {
        console.log(OutputPrefix.TOTAL_COLLECTION,
            this.getName(), this.getEarnings(), this.getDiscountsGiven())
        console.log(OutputPrefix.PASSENGER_TYPE_SUMMARY)
        for (let passenger of this.getPassengersList()) {
            console.log(passenger[0], passenger[1])
        }
    }
}



class StationBuilder {
    #station_pool // All stations

    constructor() {
        this.#station_pool = new Map()
    }

    addStation = function(station) {
        if (this.#station_pool.has(station.getName())) {
            return false
        }
        this.#station_pool[station.getName()] = station
        return true
    }

    getStation = function(station_name) {
        return this.#station_pool.get(station_name)
    }

    getAllStations = function() {
        let stations = new Array.from(this.#station_pool.values())
        return stations
    }
}


modules.exports = {
    StationBuilder,
    MetroStation
}